compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -377,6 +377,24 @@ object PatternMatcher {
             // Generate a throwaway but type-correct plan.
             // This plan will never execute because it'll be guarded by a `NonNullTest`.
             ResultPlan(tpd.Throw(tpd.nullLiteral))
+          else if (!extractor.symbol.exists) { // structural unapply
+            val selectable :: ctags :: implicits2 = implicits
+            val mt @ MethodType(_) = extractor.tpe.widen
+            val dynName = Literal(Constant(extractor.asInstanceOf[RefTree].name.toString))
+            val unapp =
+              selectable
+                .select(nme.applyDynamic)
```

Comment:
Could this be done in typer instead of in pattern matcher? Doing it late won't work as expected if applyDynamic is a macro for example.

Reply:
We could also not support this feature, I don't think it's very important.

Reply:
I agree, it would be nice to not support it. Doing it in typer would complicate the invariant about patterns.